apps/managedidentity/managedidentity.go

```diff
@@ -76,62 +93,155 @@ func (s systemAssignedValue) value() string { return string(s) }
+}
+
+type AcquireTokenOptions struct {
+	claims string
+}
+
+type ClientOption func(o *ClientOptions)
 
-func (w withClaimsOption) AcquireTokenOption()           {}
-func (w withHTTPClientOption) AcquireTokenOption()       {}
-func (w withHTTPClientOption) apply(opts *clientOptions) { opts.httpClient = w.HttpClient }
+type AcquireTokenOption func(o *AcquireTokenOptions)
 
 // WithClaims sets additional claims to request for the token, such as those required by conditional access policies.
 // Use this option when Azure AD returned a claims challenge for a prior request. The argument must be decoded.
 func WithClaims(claims string) AcquireTokenOption {
-	return withClaimsOption{Claims: claims}
+	return func(o *AcquireTokenOptions) {
+		o.claims = claims
+	}
 }
 
 // WithHTTPClient allows for a custom HTTP client to be set.
-func WithHTTPClient(httpClient ops.HTTPClient) Option {
-	return withHTTPClientOption{HttpClient: httpClient}
+func WithHTTPClient(httpClient ops.HTTPClient) ClientOption {
+	return func(o *ClientOptions) {
+		o.httpClient = httpClient
+	}
 }
 
 // Client to be used to acquire tokens for managed identity.
 // ID: [SystemAssigned()], [ClientID("clientID")], [ResourceID("resourceID")], [ObjectID("objectID")]
 //
 // Options: [WithHTTPClient]
-func New(id ID, options ...Option) (Client, error) {
-	fmt.Println("idType: ", id.value())
-
-	opts := clientOptions{
-		claims: "claims",
+func New(id ID, options ...ClientOption) (Client, error) {
+	opts := ClientOptions{
+		httpClient: shared.DefaultClient,
 	}
 
 	for _, option := range options {
-		option.apply(&opts)
+		option(&opts)
 	}
 
-	authInfo, err := authority.NewInfoFromAuthorityURI("authorityURI", true, false)
+	client := Client{
+		miType:     id,
+		httpClient: opts.httpClient,
+	}
+
+	return client, nil
+}
+
+func createIMDSAuthRequest(ctx context.Context, id ID, resource string, claims string) (*http.Request, error) {
+	var msiEndpoint *url.URL
+	msiEndpoint, err := url.Parse(imdsEndpoint)
 	if err != nil {
-		return Client{}, err
+		return nil, errors.New("error creating URL as parsing the URL filed")
 	}
 
-	authParams := authority.NewAuthParams(id.value(), authInfo)
-	client := Client{ // Note: Hey, don't even THINK about making Base into *Base. See "design notes" in public.go and confidential.go
-		AuthParams:      authParams,
-		cacheAccessorMu: &sync.RWMutex{},
-		// manager:         storage.New(token),
-		// pmanager:        storage.NewPartitionedManager(token),
+	msiParameters := msiEndpoint.Query()
+	msiParameters.Add(apiVersionQuerryParameterName, "2018-02-01")
+
+	resource = strings.TrimSuffix(resource, "/.default")
+
+	msiParameters.Add(resourceQuerryParameterName, resource)
+
+	if len(claims) > 0 {
+		msiParameters.Add("claims", claims)
+	}
+
+	switch t := id.(type) {
+	case ClientID:
+		if len(string(t)) > 0 {
+			msiParameters.Add(miQuerryParameterClientId, string(t))
+		} else {
+			return nil, fmt.Errorf("clientId parameter is empty for %T", t)
```

Comment:
These checks should happen in the constructor (the New function)? Otherwise we will repeat them for each new source we add (Arc, App Service etc.)

Reply:
Moved this to `New()` for `Client`
Added test to check that